RFC: Serialization and deserialization (#1944)
awslabs/smithy-rs: design/src/rfcs/rfc0023_serialization_and_deserialization.md

RFC: Serialization and Deserialization
=============

<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
> Status: RFC
>
> Applies to: Builder data types, data types that builder types may hold as a member as well as three data types `DateTime`, `Document`, `Blob` and `Number` implemented in `aws_smithy_types` crate.

# Terminology
- Builder  
 Refers to data types prefixed with `Builder`, which converts itself into a corresponding data type upon being built. e.g. `aws_sdk_dynamodb::input::PutItemInput`.
- serde  
  Refers to `serde` crate.
- `Serialize`   
  Refers to `Serialize` trait avaialble on `serde` crate.
- `Deserialize`  
  Refers to `Deserialize` trait available on `serde` crate.
  

# Overview
We are going to implement Serialize and Deserialize traits from serde crate on some data types.   
Data types are, a) builder data types and b) data types that builder types may have on its field(s) and c) `DateTime`, `Document`, `Blob` and `Number` implemented in `aws_smithy_types` crate.

`DateTime` and `Blob` implements different serialization/deserialization format for human-readable and non-human readable format.

> Additionally, we add two functions, `send_with`  and `replace_parameter` to the fluent builder to allow users to use the data they deserialized in the operation.

Author: ok i will fix it up once we decide what name we should be giving.

<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
RFC: Serialization and Deserialization
=============

<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
> Status: RFC
>
> Applies to: Builder data types, data types that builder types may hold as a member as well as three data types `DateTime`, `Document`, `Blob` and `Number` implemented in `aws_smithy_types` crate.

# Terminology
- Builder  
 Refers to data types prefixed with `Builder`, which converts itself into a corresponding data type upon being built. e.g. `aws_sdk_dynamodb::input::PutItemInput`.
- serde  
  Refers to `serde` crate.
- `Serialize`   
  Refers to `Serialize` trait avaialble on `serde` crate.
- `Deserialize`  
  Refers to `Deserialize` trait available on `serde` crate.
  

# Overview
We are going to implement Serialize and Deserialize traits from serde crate on some data types.   

> Data types are, a) builder data types and b) data types that builder types may have on its field(s) and c) `DateTime`, `Document`, `Blob` and `Number` implemented in `aws_smithy_types` crate.

Author: fixed.

> `DateTime` and `Blob` implements different serialization/deserialization format for human-readable and non-human readable format.

Author: I added some notes there.

Encoding must be carried out by `base64::encode` function available from `aws_smithy_types` crate.  
Non-human readable format serializes `Blob` with `fn serialize_bytes`.

- Reason behind the implementation of human-readable format
 
`aws_smithy_types` crate comes with functions for encoding/decoding base 64, which makes the implementation simpler.  
Additionally, AWS CLI and AWS SDK for other languages require data to be encoded in base 64 when it requires `Blob` type as input.  

We also considered serializing them with `serialize_bytes`, without encoding them with `serialize_bytes`.
In this case, the implementation will depend on the implementation of the library author.

There are many different crates, so we decided to survey how some of the most popular crates implement this feature.

| library    | version | implementation  | all time downloads on crate.io as of writing (Dec, 2022) |
| ---------- | ------- | --------------- | -------------------------------------------------------- |
| serde_json | 1.0     | Array of number | 109,491,713                                              |

> | tom        | 0.5.9   | Array of number | 63,601,994                                               |

Author: let me fix it.

For serialization, `DateTime` format already implements a function to encode itself into RFC-3339 format.   
For deserialization, it is possible to accept other formats, we can add this later if we find it reasonable.   

- Reason behind the implementation of a non-human readable format

Serializing them as tuples of two integers results in a smaller data size and requires less computing power than any string-based format.  
Tuple will be smaller in size as it does not require tagging like in maps.  

### Document
`Serialize` and `Deserialize` is implemented with derive macro.
Additionally, it implements container attribute `#[serde(untagged)]`.  
Serde can distinguish each variant without tagging thanks to the difference in each variant's datatypes.   

### Number
`Serialize` and `Deserialize` is implemented with derive macro.
Additionally, it implements container attribute `#[serde(untagged)]`.  

Serde can distinguish each variant without a tag as each variant's content is different.   

## Builder Types and Non-Builder Types
Builder types and non Builder types implement `Serialize` and `Deserialize` with derive macro.

> We considered implementing manually or implementing them as part of code-gen; We are not able to find any advantage that justifies the complexity.

Author: gotcha.

We considered removing traits from these data types, however, the code-gen framework does not carry the necessary metadata to determine whether the data is the builder type of an output type or not.
We conclude that we must avoid such a technical challenge to bring this RFC to life.

## `fn set_fields` to allow users supply externally created `Input`s

SDK does not have a method that allows users supply deserialized inputs.
Thus, we add a new method `fn set_fields` to `Client` types.
This method accepts inputs and replace all parameters that `Client` has with the new one.

```rust
pub fn set_fields(mut self, builder: path::to::builder_type) -> path::to::builder_type {
    self.inner = new_parameter;
    self
}
```

# Other Concerns
## Model evolution 
SDK will introduce new fields and it is possible that we may see new data types in future.  

We believe that this will not be a problem.

### Introduction of New Fields
Most fields are `Option<T>` type. 
When user de-serializes data written for a format before the new fields were introduced, new field will be assigned with `None` type.

> If a field isn't `Option`, `serde` uses `Default` trait to generate data to fill the field.  

Reviewer: Ah, OK.

> If a field isn't `Option`, `serde` uses `Default` trait to generate data to fill the field.  

Author: I will mention this on the terminology section.
Thanks for pointing it out!

> We believe that we could 

Author: Thank you for pointing it out. I will fix it.

> If the new field is not an `Option<T>` type and has no `Default` implementation, we must implement a custom de-serialization logic.

Author: At this point in time, these two are the only ones without Default trait.

- `aws_smithy_http::event_stream::Receiver`
- `aws_smithy_http::event_stream::EventStreamSender`

Number of data types could grow in future.

these data types seems to be an exception though.  
these data types are expected to work with a server, and you can't really have a `deafult server

I'm not quite sure

> If the new field is not an `Option<T>` type and has no `Default` implementation, we must implement a custom de-serialization logic.

Author: I checked transcribe crate but it seems like there are only 12 or so.

> If the new field is not an `Option<T>` type and has no `Default` implementation, we must implement a custom de-serialization logic.

Author: Sorry I think things I wrote here is not well phrased. Let me fix it up.

In case of serilization, introduction of new fields will not be an issue unless the data format requires a schema. (e.g. parquet, avro) However, this is outside of our responsibility.

## Introduction of New Data Type
If new field introduces a new data type, it will not require any additional work if the data type can derive `serde` traits.  

If the data cannot derive `serde` traits on it's own, then we have two options.
To clarify, this is the same approach we took on `Data Type to skip` section.
1. skip

>    We will simply skip serializing/de-serializing. However, we may need to implement custom serialization/de-serialization logic if a value is not wrapped with `Option`.

Author: gotcha

  // We can implement a custom serialization logic
  #[serde(serialize_with = "custom_serilization_logic", deserialize_with = "custom_deserilization_logic")]
  not_derive_able: SomeComplexDataType,
  // Serialization will be skipped, and de-serialization will be handled with the function provided on default tag
  #[serde(skip, default = "default_value")]
  skip_with_custom: DataTypeWithoutDefaultTrait,
}
```
# Discussions

## Serialization and de-serialization support for an entire response/request
The problem with serialization/de-serialization of an entire response/request the lack of data type that can be mapped to `serde`'s data model field by field.

Currently, SDK has no data type that represents an entire response or request that can be mapped to `serde`'s data model; Thus, you must introduce a schema and implement logics that allows users to serialize/de-serialize their data.

Although this RFC does not solve this issue, we believe that this RFC will help future contirbutor who wishes to implement serialization and de-serialization support for an entire response/request.

## Sensitive Information
If serialized data contains sensitive information, it will not be masked.  

> Users must be cautious to avoid compromisation

Reviewer: There is a way. You can check the member shape or it's containing shape for the `SensitiveTrait`. The code would look similar to this:

```kt
/**
 * If the root shape is sensitive, then all member shapes are considered sensitive. Otherwise, iterate over member
 * shapes and filter out any missing the `SensitiveTrait`.
 */
private fun findShapeSensitivity(rootShape: Shape): List<MemberShape> = if (rootShape.hasTrait<SensitiveTrait>()) {
    rootShape.members().toList()
} else {
    rootShape
        .members()
        .filter { it.hasTrait<SensitiveTrait>() }
}
```

> Users must be cautious to avoid compromisation

Author: That's cool. thanks!

| cargo build --release --features unstable-serde-deserialize | 1m0.107s  | 5m10.231s | 0m11.699s |
| cargo build --all-features                                  | 0m48.959s | 2m45.688s | 0m13.359s |
| cargo build --release --all-features                        | 1m3.198s  | 5m26.076s | 0m12.311s |

- `aws-sdk-ec2`
 
| command                                                     | real time | user time  | sys time  |
| ----------------------------------------------------------- | --------- | ---------- | --------- |
| cargo build                                                 | 1m20.041s | 2m14.592s  | 0m6.611s  |
| cargo build --release                                       | 2m29.480s | 9m19.530s  | 0m15.957s |
| cargo build --features unstable-serde-serialize             | 2m0.555s  | 4m24.881s  | 0m16.131s |
| cargo build --release --features unstable-serde-serialize   | 2m45.002s | 9m43.098s  | 0m16.886s |
| cargo build --features unstable-serde-deserialize           | 3m10.857s | 5m34.246s  | 0m18.844s |
| cargo build --release --features unstable-serde-deserialize | 3m47.531s | 10m52.017s | 0m18.404s |

> | cargo build --all-features                                  | 3m31.473s | 6m1.052s   | 0m19.681s |

Author: can we create DLL with rust?

> | cargo build --all-features                                  | 3m31.473s | 6m1.052s   | 0m19.681s |

Reviewer: We'd reorganize all the SDK crates so that everything related to an operation could be activated/deactivated with a single feature. Users would decide which operations they wanted to use and then add the names of those operations as feature. We have a [separate RFC describing it in detail](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0026_client_crate_organization.md).

> | cargo build --all-features                                  | 3m31.473s | 6m1.052s   | 0m19.681s |

Author: Cool, thanks!

# Feature Gate
`Serde` traits are implemented behind feature gates. 
`Serialize`  is implemented behind `serialize`., while `Deserialize` is implemented behind  `deserialize`.
Users must enable `unstable` feature to expose those features.

We considered giving each feature a dedicated feature gate such as `unstable-serde-serialize`. 
In this case, we will need to change the name of feature gates entirely once it leaves the unstable status which will cause users to make changes to their code base.
We conclude that this brings no benefit to the users.

## Keeping both features behind the same feature gate
We considered keeping both features behind the same feature gate.  
There is no significant difference in the complexity of implementation.
We do not see any benefit in keeping them behind a same feature-gate as this will only result in increase of compile time when users are not in need of one of the feature.

## Different feature gates for different data types
We considered implementing different feature gates for output, input and their corresponding data types.
For example, output and input types can have `output-serde-*` and `input-serde-*`.

> We are unable to do this as relevant meta data is not available during the code-gen.

Reviewer: We may be able to do this in a later feature anyways, so I think it's totally fine to avoid defining this.

<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
RFC: Serialization and Deserialization
=============

<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
> Status: RFC
>
> Applies to: builder structs and other data types that are connected to builder structs

# Overview
There has been some github issues requesting for the implementation of Serialize/Deserialize traits provided serde crate.  
At the current state of sdk, user must implement their own solution in order to serialize/deserialize sdk's datatypes.  
This is a RFC for implementing serde traits.

# Potential Use Cases
- Testing
- Reusable conifiguration, infrastructure as a service

> - Logging

Reviewer: - I think I understand the testing use case.
- For reusable configuration and infrastructure as a service, I'd like to see more specifics on how this change would accomplish that.
- For logging, I'd like to see a comparison with our current logging practices.
  - (Random thought that just occurred to me) We redact sensitive information in logs. We don't want to encourage people to use this feature for logging unless we can still do that or users understand how this affects their security posture.

Author: > For reusable configuration and infrastructure as a service, I'd like to see more specifics on how this change would accomplish that.
> For logging, I'd like to see a comparison with our current logging practices. 

Well, I've been creating a new struct that implements serde traits and converts them into builder types.
For example;
```rust
#[derive(Default, serde::Serialize, Deserialize)]
struct Config {
    job_definition: String,
    job_queue: String,
    job_timeout: i32,
}

impl Config {
  fn job_timeout(&self) -> JobTimeout {
    JobTimeout::builder()
        .attempt_duration_seconds(self.job_timeout)
        .build()
  }
  fn name(&self) -> String {
    format!("{}-{}-{}", create_uuid(), date(), name_of_host_computer());
  }
  fn to_batch(&self) {
    let shared_config = aws_config::from_env().load().await;

    let batch_client = aws_sdk_batch::Client::new(&shared_config);
    let res = batch_client
        .submit_job()
        .job_name(self.name().as_str())
        .job_definition(self.job_definition.as_str())
        .job_queue(self.job_queue.as_str())
        .timeout(self.job_timeout())
        .send()
        .await
        .unwrap();

  }
}
```

This is an actual piece of code that comes from one of my tools.

I found this pretty useful since I can just write stuff in rust and sometimes dynamically generate some parameters.  
The downside is that I need to create a struct that converts itself into SDK's builder types.

For logging, I was thinking of a situation where you want to reproduce something;   
As far as I'm concerned, the current logging services that AWS offers are not exactly designed for that purpose.   

I didn't put any thoughts about the sensitive info;  
I do think that there are security concerns, but I think this is something that programmers should manage.  
Even without `serde` all those data can be extracted and stored somewhere.

Maybe we might want to mention potential danger in the document.

Reviewer: At AWS, we have a concept we call "working backwards." We think about our customer's needs, and then come up with solutions to meet that need. This RFC currently reads a bit more like a solution in search of a problem. I think that this RFC would be improved by starting with the customer issues that it solves, including some example code of how customers will use your new feature to solve the problem.

There are multiple solutions to most problems, and by focusing on one, we do a disservice to customers. In your specific case, we already have several customer issues that you linked in the description of this PR. I think it would be good to demonstrate, in the RFC itself, how your proposal would meet the needs of those customers.

Also, while you did mention the tradeoffs of your proposal, that information would be even better contextualized if you could directly link it to those customer needs. Customers don't want "a cure that's worse than the disease", but they may be willing to increase compile times in CI if it makes writing and understanding tests more convenient.

> I do think that there are security concerns, but I think this is something that programmers should manage.

We also want things to be "secure by default." If doing something has security implications, we'll need to do our best to make that clear before customers start doing it. It's less about making the SDK completely safe to use and more about making it easy for customers to understand the risk of taking an action.

I know that's a lot to take in, I hope it helps! Let me know if you want clarification on anything.

 
... and more!

# Implementation
Add `serde::Serialize` and `serde::Deserialize` to builder datatypes.

# Discussion
## Enum Representation 
serde allows programmers to use one of four different tagging ([internal, external, adjacent and untagged](https://serde.rs/enum-representations.html)) when serializing enum.  

> I suggest going with `untagged` variant.  

Reviewer: If we go with `untagged` it might make it hard (or impossible) to differentiate between models that are very similar. Given that, it would be nice to see an example of what serialization and deserialization of something like a DynamoDB [PutItem](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/client/fluent_builders/struct.PutItem.html) operation would look like from a future user's perspective.

Reviewer: _(I don't mean to veto this, I'm just curious how it would work.)_

Author: I think I completely messed up here. 

Initially, I thought about using adjacent tagging, but on the last minute before submitting this RFC I changed it to untagged because I mistakenly believed that every enums that you can find in the SDKs looked something like this. 
```rust
// https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/account/src/model.rs
pub enum AlternateContactType {
    #[allow(missing_docs)] // documentation missing in model
    Billing,
    #[allow(missing_docs)] // documentation missing in model
    Operations,
    #[allow(missing_docs)] // documentation missing in model
    Security,
    /// Unknown contains new variants that have been added since this code was generated.
    Unknown(String),
}
```

Author: I think I want to replace this section with something like this.

# Discussion
## Enum Representation 
serde allows programmers to use one of four different tagging ([internal, external, adjacent and untagged](https://serde.rs/enum-representations.html)) when serializing enum.  

- untagged  
  You cannot deserialize serialized data in some cases. 
  For example, [aws_sdk_dynamodb::model::AttributeValue](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/model/enum.AttributeValue.html) has `Null(bool)` and `Bool(bool)`, which you cannot distinguish serialized values without a tag.
  
- internal  
  This results in compile time error.
  [*Using a #[serde(tag = "...")] attribute on an enum containing a tuple variant is an error at compile time*](https://serde.rs/enum-representations.html).

Author: Only question that remains is whether you want to use external tagging or adjacent tagging.  

Both of them will work.  

I think we should go with external tagging, since this is the default way and there is less question to answer.

If you want to go with adjacent tagging, you need to figure out what the name of the field is going to be.

Here is an example from the serde's document; You need to tell serde what the name of the field for tag/content is going to be.
```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
enum Block {
    Para(Vec<Inline>),
    Str(String),
}
```

> This way serialized data can be used with aws-cli's file-based input.  

Reviewer: Is this input specified anywhere? It would be good to reference it in the RFC so that the future implementer (probably you) can write tests ensuring it works.

Author: I messed it up.

I mistakenly believed that AWS CLI's file based input is automatically generated from smithy.

I couldn't find anything that backs me up.

Let me delete this section.

Reviewer: I think compatibility with the CLI would still be cool, maybe we could do this with a later feature that could read the same format? In cases like this, it's nice to say that we want this to be possible but that we leave it to some later RFC to define how.

## Backwards Compatability
- When new fields are added to builder  

> Fields of builder structs are always Option\<T>.

Reviewer: When referring to Rust types, using backticks is preferred:
```suggestion
Fields of builder structs are always `Option<T>`.
```

When serde is unable to find a field when deserializing the data, it assigns `None` to the corresponding field.  

> It is highly unlikely that deserialization of data to fail when new fields are introduced to builder types.

Reviewer: Instead of saying that it's unlikely to fail, I think it would be more helpful to outline what would cause it to fail and how we'd address those issues. If the issues can't be addressed, then we should define our strategy for informing users of the limitations of our serde implementation.

Author: OK. let me fix it up.

Author: What do you think about something like;

When serde is unable to find a field when deserializing the data, it assigns `None` to the corresponding field.    
Thus, deserialization will not fail.

Users must be aware that data deserialized with older version of SDK may not have the field that the newer version has.

Reviewer: That strategy sounds good. You could even add code examples to show what this would look like from a user's perspective.

- When fields are removed from builder   

> Users will not be able to access that removed fields via builder, however, users should be able to access it via different datatype. (e.g. `serde_json::Map`)

Reviewer: We can't ever remove a field from a builder because that would be a breaking change.

Author: Thank you for letting me know. 
I will delete this section.

- feature gate for deriving `serde::Serialize` and `serde::Deserialize`
We are going to implement them behind a feature-gate.  
The name of the feature gate is going to be `unstable-serde-de` and `unstable-serde-ser`.
```rust
#[cfg_attr(feature = "unstable-serde-ser"), derive(serde::Serialize)]
#[cfg_attr(feature = "unstable-serde-de"), derive(serde::Deserialize)]
```

> This allow users to have better control of their compile time when compared to implementing them both on behind a same feature-gate.

Reviewer: There are a few grammar issues in this RFC. Would you mind running it through a spelling/grammar checker? I think it would improve the readability.

As an example, [this tool](https://writer.com/grammar-checker/) suggests updating the sentence to read:
```suggestion
This allows users to have better control of their compile time when compared to implementing them both behind the same feature-gate.
```

Author: Let me fix it up. Thank you for bringing this up.

## implementing `serde::Serialize` and `serde::Deserialize` on builder types

> I am going to implement serde traits on existing builder types, instead of introducing new data types for serialization/deserialization.

Reviewer: Instead of framing this RFC as "What I would/will do", could you frame it as "we should do this". In this case, you'll almost certainly be the person implementing this, but the RFC should be a proposal of what "we" (everyone contributing to this project) should do.

Author: OK

Things that I considered
- Introduction of new data types will be more complicated; This will require an implementation of a function which converts the new data type into an corresponding data type.
- Compile time
Introduction of new data type should result in longer compile time than the other way around.
This one is based on my unscientific assumption; No testing has been done to verify this.
- Binary size

> Binary size should become bigger, however, it should be managible.

Reviewer: I'd like to see more detail here. You could run some informal tests yourself or perhaps reference things that other have written on this subject.

Author: OK. Let me figure this out.

Author: What do you think about running a benchmark once I finish implementing it?

Reviewer: If you find an article with a benchmark, you could use their code and run it yourself to see if their assertions are true.

If you can't find data on the specific costs, then that's fine so long as we can be confident that users that aren't using this feature are paying the cost for it.

Reviewer: Phrased another way: "Only people using this feature will pay the cost of increased compile times."

 

## Testing
Ideally, we want every data format to work with the sdk, however,
- there are large number of data format out there and not every one of them are implemented for serde
- even when serde implementations are available, some of them require programmers to pass additional pieces of data (e.g. [apache-avro](https://docs.rs/apache-avro/latest/apache_avro/) requires programmers to provide schema definition) which requires considerable investment in time to test out
- You can't simply map rust lang's data in some cases e.g. XML, CSV.  
  To clearify, it is technicalliy possible to implement something that works with XML and CSV, but there are things to figure out, for example
    - in case of CSV, you need to figure out how to handle nested structures

>     - in case of XML, you need to figure out what to do with attributes, child tags, contents ... etc

Reviewer: I think it's great that you're considering these tradeoffs.

Author: thanks :D

Having said, I suggest starting with following data format.
- JSON
- YAML

> - TOML

Reviewer: I think it might be easier to focus on supporting one initially and then adding support for others later based on user interest

Author: I will go with JSON then.

It should be the format that most AWS users are familiar with so.

Reviewer: Sounds reasonable to me

  
Those data formats are popular and it puts strong emphasis on human-readability.  

Since we could use serialized data used in testing for examples for new users, I think this is a fair start.  

> Also, many popular data format should work if it is going to work with these 3. (I have no scientific basis for this. It's just my gut feeling.)

Reviewer: I think this statement is too vague. It would be more useful to have some specifics here. Why would other data formats work?

Author: I think we want to replace this section entirely with something like this.
```
Having said that, we will create tests to ensure that serialization/deserialization will always work the JSON format.  
AWS has services that makes extensive use of JSON such as JSON policy document, and we believe that many AWS users are familiar with JSON format already and even when a user is not familiar with it, it will not be difficult for a user to find a person for assistance.
  
```

Author: That definitely sounds better.

Thank you!

Changes checklist
-----------------

- [ ] Add `#[cfg_attr(feature = "unstable-serde-ser"), derive(serde::Serialize)]` to builder types
- [ ] Add `#[cfg_attr(feature = "unstable-serde-de"), derive(serde::Deserialize)]` to builder types
- [ ] Implement tests for above mentioned data format

> - [ ] Prepare files/string for testing deserialization/serialization

Reviewer: I think we should also add a line for creating an example in our docs repo that explains how to use this feature in testing.

Author: OK.

> <!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->

Reviewer: Feel free to remove these comments if you like.
